crates/circuit/src/dag_circuit.rs

```diff
@@ -6972,6 +6981,47 @@ impl DAGCircuit {
         Ok(new_node)
     }
 
+    /// Substitute an operation in a node with a new one. The wire counts must match and the same
+    /// argument order will be used.
+    pub fn substitute_op(
+        &mut self,
+        node_index: NodeIndex,
+        new_op: PackedOperation,
+        params: SmallVec<[Param; 3]>,
+        extra_attrs: ExtraInstructionAttributes,
+    ) -> PyResult<()> {
+        let old_packed = self.dag[node_index].unwrap_operation();
+        let op_name = old_packed.op.name().to_string();
+
+        if old_packed.op.num_qubits() != new_op.num_qubits()
+            || old_packed.op.num_clbits() != new_op.num_clbits()
+        {
+            return Err(DAGCircuitError::new_err(
+                format!(
+                    "Cannot replace node of width ({} qubits, {} clbits) with operation of mismatched width ({} qubits, {} clbits)",
+                    old_packed.op.num_qubits(), old_packed.op.num_clbits(), new_op.num_qubits(), new_op.num_clbits()
+                )));
+        }
+        let new_op_name = new_op.name().to_string();
+        let new_weight = NodeType::Operation(PackedInstruction {
+            op: new_op,
+            qubits: old_packed.qubits,
+            clbits: old_packed.clbits,
+            params: (!params.is_empty()).then(|| params.into()),
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op: OnceLock::new(),
+        });
+        if let Some(weight) = self.dag.node_weight_mut(node_index) {
+            *weight = new_weight;
+        }
+
+        // Update self.op_names
+        self.decrement_op(op_name.as_str());
```

Comment:
This looks like it might do a roundtrip `&str -> String -> &str`?

Reply:
I did just copy what was there before I split this out to a method for working with PackedInstruction directly. I'll play with it a little bit and see if I can remove the double conversion (I have a feeling it's acting as a copy to avoid a borrow checker failure).

Reply:
Yeah I just looked at it the `to_string()` is to copy the string. There is a shared reference issue if a copy isn't done here because everything is based on a references to dag and we're mutating it. So we need to copy the name string before we mutate. But the `decrement_op()` and `increment_op()` methods require `&str` as the input type.